### PR TITLE
build: update vault/sdk to a valid version

### PR DESCRIPTION
**Pull request by thotz**
Build of rook images was failing with following errors
go: github.com/libopenstorage/secrets@v0.0.0-20201006135900-af310b01fe47 requires
	github.com/hashicorp/vault@v1.4.2 requires
	github.com/hashicorp/go-kms-wrapping@v0.5.1 requires
	github.com/hashicorp/vault/sdk@v0.1.14-0.20191229212425-c478d00be0d6: invalid version: unknown revision c478d00be0d6
make: *** [build/makelib/golang.mk:166: go.mod.update] Error 1

Updating version of vault/sdk in go.mod to resolve this issue

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Under what circumstances was the build failing? Which go version? Is this failing in CI builds or only for you locally?

**Comment by thotz** [comment]
> Under what circumstances was the build failing? Which go version? Is this failing in CI builds or only for you locally?

I have seen failures in two scenarios locally in my systems:

* Updating my existing repo
* cloning new repo 

Please refer https://github.com/rook/rook/issues/6583 for details

**Review by leseb — CHANGES_REQUESTED** [review]
Before doing any changes, I'd like to understand why the CI can successfully build on a fresh repo. Right now, I don't see why we would change anything.

Have you tried purging your `vendor` dir and rebuild a new one?

**Comment by thotz** [comment]
> Before doing any changes, I'd like to understand why the CI can successfully build on a fresh repo. Right now, I don't see why we would change anything.
> 
> Have you tried purging your `vendor` dir and rebuild a new one?

I had one fedora 31(go1.13.6) machine in which I updated the code with latest, then tried to build which got failed.

Then I tried on fedora 32(go1.14.10) machine (freshly set up) and tried to build rook in that which also failed.

I didn't clean up my vendor directory.

**Comment by BlaineEXE** [comment]
I can build locally without setting GOPROXY, and CI can also build. This seems like an environment issue more than a modules issue. It doesn't seem like a bad PR to merge, but it also doesn't seem critical. Maybe we should not backport this to v1.5 so we don't risk any unforeseen issues.

**Comment by raghavendra-talur** [comment]
> > Before doing any changes, I'd like to understand why the CI can successfully build on a fresh repo. Right now, I don't see why we would change anything.
> > Have you tried purging your `vendor` dir and rebuild a new one?
> 
> I had one fedora 31(go1.13.6) machine in which I updated the code with latest, then tried to build which got failed.
> 
> Then I tried on fedora 32(go1.14.10) machine (freshly set up) and tried to build rook in that which also failed.
> 
> I didn't clean up my vendor directory.

I hit a similar issue a while ago when working on rook for the first time. I can explain why some folks have a problem with the build and others don't. Fedora chose to have a non-default value(direct) for GOPROXY. That means, anyone who clones rook repo on a fedora machine will hit this issue.

Now, in this particular case, the problem is that the commit of vault sdk that the go.mod refers to is deleted in the vault repo. One of the arguments *for* using a GOPROXY is to protect projects from such commit deletions in their dependency repos. When you are using a GOPROXY you will still be able to refer to a commit which is cached in the proxy but deleted in the source repo. 

In my case, I hit this issue because I was using a GOPROXY(goproxy.io) which did not have this commit of vault sdk cached; so I hit the same error as @thotz . 

IMO, continuing to use a commit that is deleted in the source repo is not such a good idea. So, we should probably make this change in the master branch. I don't think a backport is required though.

**Comment by thotz** [comment]
> > > Before doing any changes, I'd like to understand why the CI can successfully build on a fresh repo. Right now, I don't see why we would change anything.
> > > Have you tried purging your `vendor` dir and rebuild a new one?
> > 
> > 
> > I had one fedora 31(go1.13.6) machine in which I updated the code with latest, then tried to build which got failed.
> > Then I tried on fedora 32(go1.14.10) machine (freshly set up) and tried to build rook in that which also failed.
> > I didn't clean up my vendor directory.
> 
> I hit a similar issue a while ago when working on rook for the first time. I can explain why some folks have a problem with the build and others don't. Fedora chose to have a non-default value(direct) for GOPROXY. That means, anyone who clones rook repo on a fedora machine will hit this issue.
> 
> Now, in this particular case, the problem is that the commit of vault sdk that the go.mod refers to is deleted in the vault repo. One of the arguments _for_ using a GOPROXY is to protect projects from such commit deletions in their dependency repos. When you are using a GOPROXY you will still be able to refer to a commit which is cached in the proxy but deleted in the source repo.
> 
> In my case, I hit this issue because I was using a GOPROXY(goproxy.io) which did not have this commit of vault sdk cached; so I hit the same error as @thotz .
> 
> IMO, continuing to use a commit that is deleted in the source repo is not such a good idea. So, we should probably make this change in the master branch. I don't think a backport is required though.

Thanks @raghavendra-talur for the explanation, sorry I missed it earlier. Again I hit the problem with upgraded the fedora VM?
Can u please take this change @leseb @BlaineEXE @travisn? If yes I rebase the PR.

File: go.mod [hunk 1]
```diff
@@ -52,6 +51,7 @@ require (
 // This looks "horrible", but is due to the Rook including k8s.io/kubernetes directly which is not recommended,
 // ee https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-505725449
 replace (
+	github.com/hashicorp/vault/sdk => github.com/hashicorp/vault/sdk v0.1.14-0.20201110190005-0340b40575ad
```
Comment (travisn) [inline go.mod:54]: Why do we need to pin it to this specific version? I thought if we updated the modules, it should resolve the latest automatically.

**Review by travisn — COMMENTED** [review]
Did you update this with `make go.mod.update`? I would hope that would take care of resolving valid references.

File: go.mod (hunk 1 above)
Comment (raghavendra-talur) [reply go.mod:54]: IIUC, vault/sdk is not a directly inherited module in rook. It is coming in as a dependency of github.com/libopenstorage/secrets. 

As per https://research.swtch.com/vgo-principles the recommended way to fix this error is a two step process:
1. Bump up the version of 'github.com/hashicorp/vault' in 'github.com/libopenstorage/secrets' such that it does not reference the deleted commit anymore.
2. Bump up the version of 'github.com/libopenstorage/secrets' in Rook to the commit which performed step 1.

In cases where changes like in step 1 aren't possible(If the repo is not well maintained etc), we can use the replace section of go.mod to override the version like in this PR.

When I first looked at this PR, I missed that libopenstorage/secrets is also maintained by Rook developers. I suggest using the prescribed method and not the replace option.

Comment (thotz) [reply go.mod:54]: I thought libopenstorage/secrets maintained by protwox folks, not by rook community

**Comment by alimaredia** [comment]
I've hit this same build issue on Fedora 33. If I make the changes that specify the vault version in `go.mod`, the issues go away.

**Comment by serajam** [comment]
Same here, ubuntu 21.04

**Comment by alimaredia** [comment]
This issue went away for me when I switched from the `golang` package installed via `dnf` on Fedora 33 to downloading the `.tar.gz` with a workflow similar to this: https://golang.org/doc/install

**Comment by leseb** [comment]
@thotz can you close this?

**Pull request closed without merging**